src/marqo/s2_inference/s2_inference.py

```diff
@@ -81,17 +86,98 @@ def _update_available_models(model_cache_key: str, model_name: str, validated_mo
+                                                 "time_stamp": time_stamp,
+                                                 "model_size": model_size}
+            logger.info(f'loaded {model_name} on device {device} with normalization={normalize_embeddings} at time={time_stamp}.')
         except:
             raise ModelLoadError(
                 f"Unable to load model={model_name} on device={device} with normalization={normalize_embeddings}. "
                 f"If you are trying to load a custom model, "
                 f"please check that model_properties={validated_model_properties} is correct "
                 f"and the model has valid access permission. ")
+    else:
+        time_stamp = datetime.datetime.now()
+        logger.debug(f'renew {model_name} on device {device} with new time={time_stamp}.')
+        available_models[model_cache_key]["time_stamp"] = time_stamp
+
+
+def get_model_size(model_name:str, model_properties:dict) -> (int, float):
+    '''
+    Return the model size for given model
+    Note that the priorities are size_in_properties -> model_name -> model_type
+    '''
+    if "model_size" in model_properties:
+        return model_properties["model_size"]
+
+    name_info = (model_name + model_properties.get("name", "")).lower().replace("/", "-")
+    print(name_info)
+    for name, size in constants.MODEL_NAME_SIZE_MAPPING.items():
+        if name in name_info:
+            return size
+
+    type = model_properties["type"]
+    return constants.MODEL_TYPE_SIZE_MAPPING.get(type, constants.DEFAULT_MODEL_SIZE)
+
+
+def device_memory_manage(model_name: str, model_properties: dict, device: str) -> bool:
+    '''
+    A function to manage the memory usage in devices when we want to load a new model
+    Args:
+        model_name: The name of the model to load
+        model_properties: The model properties of the model
+        device: The target device to laod the model
+    Returns:
+        True we have enough space for the model
+        Raise an error and return False if we can't find enough space for the model.
+    '''
+    model_size = get_model_size(model_name, model_properties)
+    if check_device_memory_status(device, model_size):
+        return True
+    else:
+        model_cache_key_in_device = [key for key in list(available_models) if key.endswith(device)]
+        sorted_key_in_device = sorted(model_cache_key_in_device,
+                                      key=lambda x: available_models[x]["time_stamp"])
+        for key in sorted_key_in_device:
+            logger.info(f"Eject model = `{key.split('||')[0]}` with size = `{available_models[key].get('model_size', constants.DEFAULT_MODEL_SIZE)}` from device = `{device}` "
+                        f"to save space for model = `{model_name}`.")
+            del available_models[key]
+            if check_device_memory_status(device, model_size):
+                return True
+
+        if check_device_memory_status(device, model_size) is False:
```

Comment:
Perhaps we can calculate this before ejecting all the models? So that the user doesn't have to wait around to reload their previous models

Reply:
Can we calculate this before ejecting the models, so we don't needlessly eject models?